# Homework 1: Introduction to Machine Learning

In [16]:
import numpy as np
import pandas as pd

# Load the 2026 Car Fuel Efficiency dataset
df = pd.read_csv("../data/car_fuel_efficiency_2026.csv")
df.head()


,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


### Question 1: Pandas version

What version of Pandas did you install?


In [17]:
pd.__version__


'3.0.5'

**Answer:** `3.0.6`


### Question 2: Records count

How many records are in the dataset?
- 5000
- 9000
- 10000
- 15000


In [18]:
len(df)


10000

**Answer:** `10000`


### Question 3: Fuel types

How many fuel types are presented in the dataset?
- 1
- 2
- 3
- 4


In [19]:
print("Unique fuel types count:", df['fuel_type'].nunique())
print("Unique fuel types:", df['fuel_type'].unique())


Unique fuel types count: 3
Unique fuel types: <ArrowStringArray>
['Gasoline', 'Diesel', 'Hybrid']
Length: 3, dtype: str


**Answer:** `3` (Gasoline, Diesel, Hybrid)


### Question 4: Missing values

How many columns in the dataset have missing values?
- 0
- 1
- 2
- 3
- 4


In [20]:
print("Missing values per column:")
print(df.isna().sum())
print("\nNumber of columns with missing values:", (df.isna().sum() > 0).sum())


Missing values per column:
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

Number of columns with missing values: 2


**Answer:** `2` (`horsepower` has 877 missing values, and `acceleration` has 264 missing values)


### Question 5: Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?
- 21.2
- 31.2
- 41.2
- 51.2


In [24]:
max_mpg = df['fuel_efficiency_mpg'].max()
max_mpg


np.float64(41.2)

**Answer:** `41.2`


### Question 6: Median value of horsepower

1. Find the median value of the `horsepower` column in the dataset.
2. Next, calculate the most frequent value of the same `horsepower` column.
3. Use the `fillna` method to fill the missing values in the `horsepower` column with the most frequent value from the previous step.
4. Now, calculate the median value of `horsepower` once again.

Has it changed?
- Yes, it increased
- Yes, it decreased
- No


In [22]:
# 1. Median value before imputation
median_before = df['horsepower'].median()
print(f"Median before imputation: {median_before}")

# 2. Most frequent value (mode)
mode_hp = df['horsepower'].mode()[0]
print(f"Most frequent value (mode): {mode_hp}")

# 3. Fill missing values with the mode
df_filled = df.copy()
df_filled['horsepower'] = df_filled['horsepower'].fillna(mode_hp)

# 4. Median value after imputation
median_after = df_filled['horsepower'].median()
print(f"Median after imputation: {median_after}")

# Comparison
if median_after > median_before:
    print("Has it changed?: Yes, it increased")
elif median_after < median_before:
    print("Has it changed?: Yes, it decreased")
else:
    print("Has it changed?: No")


Median before imputation: 254.0
Most frequent value (mode): 252.0
Median after imputation: 252.0
Has it changed?: Yes, it decreased


**Answer:** `Yes, it decreased`

The median changed from **254.0** to **252.0** after filling missing values with the mode (252.0).


### Question 7: Sum of weights

1. Select all the cars from Asia
2. Select only columns `vehicle_weight` and `model_year`
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the sum of all the elements of the result?
- 0.0369
- 0.369
- 3.69
- 36.9


In [23]:
# 1. Select all the cars from Asia
# 2. Select only columns vehicle_weight and model_year
# 3. Select the first 7 values
asia_cars = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7)

# 4. Get the underlying NumPy array
X = asia_cars.to_numpy()
print("X:\n", X)

# 5. Compute matrix-matrix multiplication between X.T and X
XTX = X.T @ X
print("\nXTX:\n", XTX)

# 6. Invert XTX
XTX_inv = np.linalg.inv(XTX)
print("\nXTX_inv:\n", XTX_inv)

# 7. Create array y
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
print("\ny:\n", y)

# 8. Multiply inverse of XTX with X.T, and then multiply by y
w = XTX_inv @ X.T @ y
print("\nw:\n", w)

# 9. Sum of all elements of w
sum_w = w.sum()
print(f"\nSum of weights: {sum_w:.4f} -> rounded to 3 decimal places: {sum_w:.3f}")


X:
 [[4240 1996]
 [4310 1992]
 [4230 1997]
 [4420 1978]
 [3950 2007]
 [4750 2021]
 [4450 2019]]

XTX:
 [[131950500  60750580]
 [ 60750580  28041424]]

XTX_inv:
 [[ 2.96830537e-06 -6.43071025e-06]
 [-6.43071025e-06  1.39675281e-05]]

y:
 [1100 1300  800  900 1000 1100 1200]

w:
 [0.13644777 0.2327492 ]

Sum of weights: 0.3692 -> rounded to 3 decimal places: 0.369


**Answer:** `0.369`

The computed sum of weights is approximately **0.3692**, which matches the option **`0.369`**.
